src/lib.rs

```diff
@@ -72,19 +72,178 @@ struct PointProps<'a> {
+pub fn apply_accum(
+    out: &mut Vec<f64>,
+    weights_out: &mut Vec<f64>,
     accum: &mut Mean,
     points_a: &PointProps,
     points_b: Option<&PointProps>,
+    // TODO decide bin, partition, or something else, then unify the naming
+    // TODO should bin_edges should be a member of the AccumKernel Struct
     bin_edges: &[f64],
     /* pairwise_op, */ // probably an Enum
 ) -> Result<(), String> {
-    // TODO: check that bin_edges monotonically increases
-    // TODO: if points_b is not None, make sure there is agreement in the
-    //       n_spatial_dims attribute
-    Err(String::from("Not implemented yet!"))
+    // TODO check size of output buffers
+
+    // Check that bin_edges are monotonically increasing
+    if !bin_edges.is_sorted() {
+        return Err(String::from(
+            "bin_edges must be sorted (monotonically increasing)",
+        ));
+    }
+
+    //  if points_b is not None, make sure a and b have the same number of
+    // spatial dimensions
+    if let Some(points_b) = points_b {
+        if points_a.n_spatial_dims != points_b.n_spatial_dims {
+            return Err(String::from(
+                "points_a and points_b must have the same number of spatial dimensions",
+            ));
+        }
+    }
+
+    accum.initialize();
+    // I think this alloc is worth it? Could use a buffer?
+    let squared_bin_edges: Vec<f64> = bin_edges.iter().map(|x| x.powi(2)).collect();
+
+    // complicated case first: two arrays, we combine points with pairwise_op
+    if let Some(points_b) = points_b {
+        for i_a in 0..points_a.n_points {
+            for i_b in 0..points_b.n_points {
+                // compute the distance between the points, then the distance bin
+                let distance_squared = squared_diff_norm(
+                    points_a.positions,
+                    points_b.positions,
+                    i_a,
+                    i_b,
+                    points_a.spatial_dim_stride,
+                    points_b.spatial_dim_stride,
+                    points_a.n_spatial_dims,
+                );
+                let Some(distance_bin) = get_distance_bin(distance_squared, &squared_bin_edges)
+                else {
+                    continue;
+                };
+
+                // get the value. This is hardcoded to correspont to the velocity structure
+                // function when accumulated with Mean
+                // TODO switch on pairwise op?
+                // TODO I want to talk this design through before we roll further with it
+                // I was imagining that there would be separate AccumKernels for, e.g.
+                // structure functions and correlations, rather than using Mean with
+                // different pairwise_ops. I think it might be possible to write some
+                // nice "AccumKernel combinators" to make this clean, but maybe it would
+                // require too much magic.
```

Comment:
I think the play here is to move the entire loop into a helper function like:

```
fn apply_accum(&mut accum: Mean, ..., pairwise_fn: F) -> ()
where
    F: Fn(v_a: [&f64], v_b: [&f64], u_size, u_size, u_size) -> f64)
{
  // ...
}
```

And the we can pass in
```rust
|v_a: &[f64], v_b: &[f64], i_a: usize, i_b: usize, stride_a: usize, stride_b: usize| {
   squared_diff_norm(v_a, v_b, i_a, i_b, stride_a, stride_b, n_spatial_dims).sqrt()
};
```

(we can make simplification based upon closure capture)